src/librbd/internal.cc

```diff
@@ -1706,11 +1706,17 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
 
       C_SaferCond lock_ctx;
       ictx->exclusive_lock->request_lock(&lock_ctx);
+
+      // don't block holding lock since refresh might be required
+      ictx->owner_lock.put_read();
```

Comment:
The order for blocking writes/requests was required to prevent an rbd CLI from accepting proxied requests on behalf of the client we might just have taken the lock. The block request call only increments a counter -- the block writes call might result in a refresh if and only if there was active IO (which wouldn't be coming from the rbd CLI). 

The release lock won't trigger a refresh -- that is built into the acquire lock state machine due to a recent bug fix.

The lock will be released whenever the image is closed (or requested) -- that release at the end was just a way to force a specific event order for test cases with regard to the notify update. 

This whole routine needs to be re-written as an async state machine because it's grown to be quite unmanageable with all it needs to do.